src/providers/tpm/key_management.rs

```diff
@@ -14,12 +14,61 @@ use parsec_interface::operations::{
 };
 use parsec_interface::requests::{ProviderId, ResponseStatus, Result};
 use parsec_interface::secrecy::ExposeSecret;
-use picky_asn1_x509::{RSAPrivateKey, RSAPublicKey};
-use tss_esapi::abstraction::transient::RsaExponent;
+use picky_asn1_x509::RSAPublicKey;
+use std::convert::TryInto;
 
 const AUTH_VAL_LEN: usize = 32;
 
 impl Provider {
+    pub(super) fn get_key_ctx(&self, key_triple: &KeyTriple) -> Result<PasswordContext> {
+        // Try to deserialize into the new format
+        self.key_info_store
+            .get_key_id::<PasswordContext>(key_triple)
+            .or_else(|e| {
+                // If it failed, check if it was a deserialization error
+                if let ResponseStatus::InvalidEncoding = e {
+                    // Try to deserialize into legacy format
+                    let legacy_ctx = self
+                        .key_info_store
+                        .get_key_id::<LegacyPasswordContext>(key_triple)?;
+
+                    // Try to migrate the key context to the new format
+                    let mut esapi_context = self
+                        .esapi_context
+                        .lock()
+                        .expect("ESAPI Context lock poisoned");
+                    let password_ctx = PasswordContext {
+                        key_material: esapi_context
+                            .migrate_key_from_ctx(
+                                legacy_ctx.context,
+                                Some(
+                                    legacy_ctx
+                                        .auth_value
+                                        .clone()
+                                        .try_into()
+                                        .map_err(utils::to_response_status)?,
+                                ),
+                            )
+                            .map_err(utils::to_response_status)?,
+                        auth_value: legacy_ctx.auth_value,
+                        context: utils::empty_tpms_ctx(),
+                    };
+
+                    // Grab key attributes and replace legacy entry with new one
+                    let attributes = self.key_info_store.get_key_attributes(key_triple)?;
+                    self.key_info_store.remove_key_info(key_triple)?;
+                    let _ = self.key_info_store.insert_key_info(
```

Comment:
If we don't already, it might be worth stress-testing this pattern of removal followed by re-insertion of the same key in the on-disk KIM, because I have known this kind of thing to fail due to race conditions against the filesystem (although this is going back several years). A `patch_key_info` operation in the KIM might be preferable, but I can understand that this requires a more extensive change of the kind we're trying to avoid here. On balance, I'm happy with this as it is.

Reply:
Yeah, I've been thinking about this - I'll add a `replace_key_info` to the KIM client. The worst that can happen here is that a thread tries to get info about the key in the split second between when the attributes are deleted and when they're placed back, and doesn't find any so it returns "PsaErrorNotFound".

Reply:
Done!